### ISSUE TITLE
Unable to perform quality trimming on both ends of paired end reads

### ISSUE THREAD

**Issue by taltman** [marcelm/cutadapt#234]
I am trying to use cutadapt (v 1.12, Python 2.6.6) to perform quality trimming on both ends of an Illumina paired-end read dataset.

I have tried to scrupulously follow the manual, but I find the stated documentation doesn't seem to match the behavior of the program.

Here's how I initially tried to use the program:

```
cutadapt \
                --quality-base=64 \
                -q $assembly_quality_threshold \
                -o ${file_basename}_cutadapt-assembly_1.fastq \
                --paired-output ${file_basename}_cutadapt-assembly_2.fastq \
                ${file_basename}_trimmomatic_1P.fastq ${file_basename}_trimmomatic_2P.fastq \
                | tee cutadapt-assembly-report-paired.txt
```

... and here's the output that I get:

```
Total basepairs processed:     172876027 bp
  Read 1:      86442912 bp
  Read 2:      86433115 bp
Quality-trimmed:                 2020218 bp (1.2%)
  Read 1:       2020218 bp
  Read 2:             0 bp
  ```

So, only the first read file was being trimmed.

According to the documentation, the `-q` option is supposed to work on both reads. But then there's some flanking documentation that raises ambiguity about whether it needs some additional flags to override the default 'legacy mode'. So I tried adding the `-A XXX` option to the command line, as the above execution run suggested after warning about legacy mode being enabled. Here is the result:

```
Total basepairs processed:     172876027 bp
  Read 1:      86442912 bp
  Read 2:      86433115 bp
Quality-trimmed:                       0 bp (0.0%)
  Read 1:             0 bp
  Read 2:             0 bp
```

So, now neither set of reads are being trimmed.

Any help in trying to figure out the documentation would be greatly appreciated! Thanks in advance!

**Comment by marcelm** [comment]
You are right that you need the `-A XXX` option. Can you please verify that the warning about "legacy mode" disappears when you add that? What is the value you are using for `$assembly_quality_threshold`?

I’ve tested the above command and it does trim both reads for me. It would be most helpful if you could send me one or two read pairs so that I can try to reproduce the problem.

Note that I’ve dropped support for Python 2.6. It does still work at the moment, and it is probably not the problem here, but in the future this is probably going to change.

**Comment by taltman** [comment]
Hi Marcel,

Thanks for your reply.

Yes, to confirm, using the '-A XXX' trick does silence the legacy mode warning message.

I attempted the trimming twice, once with `-q 20` and once with `-q 0.05` (to approximate Mott's alternative algorithm for trimming as described in PHRED). Both times I observed the behavior that I listed above.

Here are two example paired-end reads from my set. Note that I used Trimmomatic to remove TruSeq adapters first, but with no quality trimming performed.

Any help that you can provide would be greatly appreciated! Thanks in advance!

```
head -n 8 *.fastq
==> s_4_1_allseq_trimmomatic_1P.fastq <==
@HWUSI-EAS466__1_4_1_1007_1251
NTGAAAAGTTCAACACGGCAGTGGCGGCGGTGGTGG
+HWUSI-EAS466__1_4_1_1007_1251
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1008_17091
NGTTCGGCAAGGGCCGTTCAACAATTATGCTTGTAA
+HWUSI-EAS466__1_4_1_1008_17091
BJJJJNQQQQ````````````````````````V`

==> s_4_1_allseq_trimmomatic_2P.fastq <==
@HWUSI-EAS466__1_4_1_1007_1251
ATCAGCGTGACCCAATTCATGCCACAATTCTTCGGT
+HWUSI-EAS466__1_4_1_1007_1251
bbbbbbbbbbbbbbbbababbb_bbbbbbbbbbbbb
@HWUSI-EAS466__1_4_1_1008_17091
CATTGCGCCGAAACGCTTCCTCTGCCAACCGGTTGA
+HWUSI-EAS466__1_4_1_1008_17091
bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbcbbb
```

**Comment by marcelm** [comment]
Thanks for the test data.

Note that using non-integer values with `-q` is not supported, so you should get an error message for `-q 0.05`.

In your example, both R2 reads end with high-quality bases (`b` encodes quality 34), so I would totally expect these reads not to be trimmed with `-q 20`. When I change some of the final `b` to `B`, then these get trimmed off. I saved your files as `in.1.fastq` and `in.2.fastq` and ran:
```
cutadapt --quality-base=64 -q 20 -A XXX -o out.1.fastq -p out.2.fastq in.1.fastq in.2.fastq
```
I get this (on the version where I replaced a few `b` with `B`):
```
Quality-trimmed:                      43 bp (29.9%)
  Read 1:            36 bp
  Read 2:             7 bp
```
Is it possible that you simply don’t have any low-quality ends in your R2 reads? Try `-q` with some higher threshold, does that change anything?

**Comment by taltman** [comment]
Hi Marcel,

Thanks for your reply!

Yes, I forgot to log transform that value. The value supplied to -q is 13.

I did a quick histogram analysis of my reads, and I don't see any
distinct differences in quality value distributions between the
forward and the reverse. I've gone ahead and found a set of examples
that show low-quality bases in the reverse end. Please let me know if
I can supply anything else that would help with debugging. Thanks in advance for your help!

Forward:
```
@HWUSI-EAS466__1_4_1_1016_17437
NGATTCTGTTTTATTTACTCTTAAAGGGGTTAATGC
+HWUSI-EAS466__1_4_1_1016_17437
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1021_9648
NACTATTTCCAATAGTTGAGTAATTGTAAAGTTGCA
+HWUSI-EAS466__1_4_1_1021_9648
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1028_13541
AACAATTTTTCCATACTTTTTGTATAAAGATCGCAG
+HWUSI-EAS466__1_4_1_1028_13541
\__ST]T[ZO]G^N^BBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1032_19603
GCTGATTCTTTCTTAAACCAAAATCCACTTAACAAA
+HWUSI-EAS466__1_4_1_1032_19603
R^Z`b`baaa^aaaO^aZ]^WaUW^^`UK``^^UX^
```

Reverse:

```
@HWUSI-EAS466__1_4_1_1016_17437
GTCCGTTAAGTTTCCACCGGAGTTATAAGAACCCGA
+HWUSI-EAS466__1_4_1_1016_17437
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1021_9648
GTTTAAGTTATTACTCTAGAAATTCGTCTTTTTGTC
+HWUSI-EAS466__1_4_1_1021_9648
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1028_13541
AAAGACCGCCGTCATCCAATCTTCCATACGCCAACG
+HWUSI-EAS466__1_4_1_1028_13541
`Y_Q``BBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
@HWUSI-EAS466__1_4_1_1032_19603
TCGAGCCTATATGGCCGGTGATCCTGGCTATAAGCG
+HWUSI-EAS466__1_4_1_1032_19603
BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
```

**Comment by marcelm** [comment]
Thanks. I've now set up an environment with Python 2.6.9 and cutadapt 1.12 and still cannot see the problem. Can you please run those two files through cutadapt with the same command that I posted above and let me know what the output is?

**Comment by taltman** [comment]
Hi Marcel,

I can confirm that it works in this test case.

I've done more debugging, and it looks like the culprit is that I try to pass some parameters to cutadapt when constructing its command line execution on my script, and I mistakenly quoted as follows:

cutadapt "-A XXX --quality-base=64" ...

The program silently accepted "XXX --quality-base=64" as a legitimate adapter string.

I think it would be helpful if cutadapt did some light sanity checks on its arguments. I think a string given as an adapter sequence that does not consist of IUPAC nucleotide base characters should trigger a warning.

Thanks for your help!

**Comment by marcelm** [comment]
Great, glad you found it!

Cutadapt already sanity-checks most of its arguments, and I try hard to give meaningful error messages. But you are right that cutadapt indeed accepts any string as an adapter. As an explanation/justification: This functionality was there in case someone wanted to trim protein sequences or any other type of textual data. It’s nice for directly working with examples such as "SEQUENCEADAPTERSOMETHINGELSE" as given in the documentation. If you put that in a FASTA file, you can just run `cutadapt -a ADAPTER in.fasta` to trim it.

I agree that the benefits are small, and this issue has shown that the costs may be large, so I’ll enforce stricter checks in the future, please see issue #235.

Thanks for helping out!

**Comment by marcelm** [comment]
I’ve implemented a stricter check, see issue #235 and commit 94afa2a
